accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+
+*   If `collection` is a set literal, then the downwards context for inference
+    of the elements of `collection` is `Set<P>` where `P` may be `?` if
+    downwards inference does not constrain the type of `collection`.
+
+    *   If `P` is `?` then the static type of `collection` is `Set<T>` where `T`
+        is the upper bound of the inferred set element types of the elements.
+
+    *   Otherwise, the static type of `collection` is `Set<T>` where `T` is
+        determined by downwards inference.
+
+    *Note that the element inference will never produce a key/value type here
+    given that downwards context.*
+
+*   If `collection` is a map literal and downwards inference has statically
+    known type `Map<K, V>` then the downwards context for the elements of
+    `collection` is `Map<K, V>`.
+
+*   If `collection` is a map literal then the downwards context for the elements
+    of `collection` is `Map<Pk, Pv>` where `Pk` and `Pv` are determined by
+    downwards inference, and may be `?` if the downwards context does not
+    constrain one or both.
+
+    *   If `Pk` is `?` then the static key type of `collection` is `K` where `K`
+        is the upper bound of the inferred key types of the elements.
+
+    *   Otherwise the static key type of `collection` is `K` where `K` is
+        determined by downwards inference.
+
+    And:
+
+    *   If `Pv` is `?` then the static value type of `collection` is `V` where
+        `V` is the upper bound of the inferred value types of the elements.
+
+    *   Otherwise the static value type of `collection` is `V` where `V` is
+        determined by downwards inference.
+
+    *Note that the element inference will never produce a set element type here
+    given this downwards context.*
+
+*   If `collection` is not syntactically known to be a set or map literal, then
+    the downwards context for the elements of `collection` is `?`, and the disambiguation
+    is done as follows:
+
+    *   If all elements `ei` have a set element type `Ti`, and at least one
+        element does not have a key/value type pair, then `collection` is a set
+        literal with static type `Set<T>` where `T` is the upper bound of `Ti`.
+        *In other words, if every element can be a set and at least one must, it
+        is a set.*
+
+    *   If all elements `ei` have key element type `Ki`, and value element type
+        `Vi` , and at least one element does not have a set element type, then
+        `e` is a map literal with static type `Map<K, V>` where `K` is the upper
+        bound of `Ki` and `V` is the upper bound of `Vi`. *In other words, if
+        every element can be a map and at least one must, it is a map.*
+
+    *   If all elements have both a set element type and a key/value type pair,
+        then the literal is ambiguous and it is a compile-time error. *This
+        occurs when all the leaf elements are dynamic spreads.*
+
```

Comment:
With the Null modification, add a case that says that if no element has a set element type, and no element has a key/value pair type, then the literal is ambiguous and it is a compile time error.  *This occurs when all the leaf elements are null aware spreads of expressions of type `Null`.